dialogflow-cx/streaming_detect_intent_infinite.py

```diff
@@ -0,0 +1,720 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This script implements a real-time bidirectional streaming audio interface
+with Google Cloud Dialogflow CX. It captures audio from the user's microphone,
+streams it to Dialogflow CX for intent detection and natural language
+processing, and plays back the synthesized audio responses from Dialogflow
+CX through the user's speakers.
+
+Dependencies:
+  - google-cloud-dialogflow-cx: Cloud Dialogflow CX API client library.
+  - termcolor: For colored terminal output.
+  - pyaudio: For interfacing with audio input/output devices.
+
+NOTE: pyaudio may have additional dependencies depending on your platform.
+
+Install dependencies using pip:
+
+```bash
+pip install -r requirements.txt
+```
+
+Before Running:
+
+  - Set up a Dialogflow CX agent and obtain the agent name.
+  - Ensure you have properly configured Google Cloud authentication
+    (e.g., using a service account key).
+
+Example usage:
+
+```bash
+python streaming_detect_intent_infinite.py projects/your-project/locations/your-location/agents/your-agent-id --language_code en-US --sample_rate 16000  --single_utterance
+```
+
+Replace `projects/your-project/locations/your-location/agents/your-agent-id`
+with your Dialogflow CX agent name.
+
+Further information is provided via:
+
+```bash
+python streaming_detect_intent_infinite.py --help
+```
+
+The script will start capturing audio from the default microphone, stream it
+to Dialogflow CX, and play back the audio responses. Press Ctrl+C to exit the
+program gracefully. Say "Hello" to trigger the Default Intent.
+"""
+
+# [START dialogflow_streaming_detect_intent_infinite]
+
+import argparse
+import asyncio
+import logging
+import os
+import signal
+import struct
+import sys
+import time
+from typing import Any, AsyncGenerator, List, Optional, Tuple
+import uuid
+
+from google.api_core import retry as retries
+from google.api_core.client_options import ClientOptions
+from google.api_core.exceptions import GoogleAPIError, ServiceUnavailable
+from google.cloud import dialogflowcx_v3
+from google.protobuf.json_format import MessageToDict
+
+import pyaudio
+from termcolor import colored
+
+# TODO: Remove once GRPC log spam is gone see https://github.com/grpc/grpc/issues/37642
+os.environ["GRPC_VERBOSITY"] = "NONE"
+
+# Configure logging
+logging.basicConfig(
+    level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s"
+)
+logger = logging.getLogger(__name__)
+
+CHUNK_SECONDS = 0.1
+
+
+def get_current_time() -> int:
+    """Return Current Time in MS."""
+    return int(round(time.time() * 1000))
+
+
+class AudioIO:
+    """Audio Input / Output"""
+
+    def __init__(
+        self,
+        rate: int,
+        chunk_size: int,
+        audio_encoding: str = "LINEAR16",
+        output_audio_encoding: str = "LINEAR16",
+    ) -> None:
+        self._rate: int = rate
```

Comment:
The purpose of a code sample is to provide a simple example on how to use the APIs or show a common pattern, not to provide a fully customizable tool. The amount of options here makes the sample look a lot more intimidating than what it actually is.

Can we inline as many of these options into where they're used? That makes the code sample easier to read, since you can directly see all the values rather than having to scroll up to see which configurations are important and which can be left to None or some other empty value.

Let's just leave here the variables needed to keep state on the object. Any configuration should be inlined.